### DISCUSSION TITLE
How do I create a custom theme?

### DISCUSSION THREAD

**Discussion by Shadowblitz16** [AvaloniaUI/Avalonia#16239] [Q&A] [answered]
I was looking at https://docs.avaloniaui.net/docs/basics/user-interface/styling/control-themes trying to figure out how to make a custom theme, But couldn't figure it out.

How would I do something like this for a button's normal state?
![panel](https://github.com/AvaloniaUI/Avalonia/assets/2453707/bec9b210-4169-48e8-a6f7-d3058e9147af)

Can I have an example in both C# and xaml?

I did manage to get a style to look like it with this...
```xml
<Styles xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">
    <Design.PreviewWith>
        <Border MinWidth ="32" MinHeight="32" BorderThickness="1,1,0,0"  BorderBrush="#d79773">
            <Border BorderThickness="0,0,1,1" BorderBrush="#4f2b13">
                <Border BorderThickness="1,1,0,0" BorderBrush="#ffe3c3">
                    <Border BorderThickness="0,0,1,1" BorderBrush="#8b573b">
                        <Border Background="#d79773">
                            <!-- Add Controls for Previewer Here -->
                        </Border>
                    </Border>
                </Border>
            </Border>
        </Border>
    </Design.PreviewWith>

    <!-- Add Styles Here -->
</Styles>
```

but I don't understand how this applies to a control and/or a specific state like normal, hovered, focused, pressed, disabled.

**Comment by timunie** [comment]
Lookup other styles in source. AwsomeAvalonia has some styling libs referenced for you to get inspired

**Reply by Shadowblitz16** [reply]
> Lookup other styles in source. AwsomeAvalonia has some styling libs referenced for you to get inspired

Those don't explain what is what though, only how it's done.
I want to understand how it works not just how to copy and paste someone else's code. 

I want to know what each property is and how the theming system works

**Reply by maxkatz6** [reply]
> I want to understand how it works not just how to copy and paste someone else's code.
> I want to know what each property is and how the theming system works

You need to ask specific questions then.
Normally, when you want to know how internals work, you have to go way deeper than most developers do. In case of Avalonia, it's a bit easier since you can read the source code or related PRs.

**Reply by maxkatz6** [reply]
In the end, if you want to create a new theme, the best way is to learn how actual themes are implemented.

**Comment by maxkatz6** [comment, marked as answer]
> I did manage to get a style to look like it with this...

`Design.PreviewWith` is not really a reusable part. It's only used for the previewer to render current XAML file.
As per that documentation page, you need to create a ControlTheme defined in the resources. And then reference it in your controls.

> How would I do something like this for a button's normal state?

Going with the same example, you need to define ControlTheme and a ControlTemplate inside of it. Template is an actual "function" that builds the control each time it's displayed. And while on it, ContentPresenter is an actual container for the "Content" your control has. Without ContentPresenter it would be just a group of borders without anything inside.
```xaml
<ResourceDictionary xmlns="https://github.com/avaloniaui"
                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">
  <Design.PreviewWith>
    <Button Content="Hello World!"
            Theme="{StaticResource CustomTheme}"
            VerticalContentAlignment="Center" />
  </Design.PreviewWith>

  <ControlTheme TargetType="{x:Type Button}"
                x:Key="CustomTheme">
    <Setter Property="MinHeight" Value="32" />
    <Setter Property="MinWidth" Value="32" />
    <Setter Property="Template">
      <ControlTemplate>
        <Border BorderThickness="1,1,0,0"  BorderBrush="#d79773">
          <Border BorderThickness="0,0,1,1" BorderBrush="#4f2b13">
            <Border BorderThickness="1,1,0,0" BorderBrush="#ffe3c3">
              <Border BorderThickness="0,0,1,1" BorderBrush="#8b573b">
                <Border Background="#d79773">
                  <ContentPresenter x:Name="PART_ContentPresenter"
                                    Content="{TemplateBinding Content}"
                                    Margin="{TemplateBinding Padding}"
                                    HorizontalContentAlignment="{TemplateBinding HorizontalContentAlignment}"
                                    VerticalContentAlignment="{TemplateBinding VerticalContentAlignment}"  />
                </Border>
              </Border>
            </Border>
          </Border>
        </Border>
      </ControlTemplate>
    </Setter>
  </ControlTheme>
</ResourceDictionary>
```

Note, I did couple of changes just to make it easier to use:
1. Extracted MinWidth/MinHeight to setters, so it can be redefined on the button itself. If it was set on the Border directly, you wouldn't be able to change that value from the outside. TemplateBinding might have help, but there is no reason to use it in this scenario.
2. Added HorizontalContentAlignment + VerticalContentAlignment, and passed down to the content presenter.

Also, see other documentation pages that are useful:
https://docs.avaloniaui.net/docs/guides/custom-controls/how-to-create-templated-controls
https://docs.avaloniaui.net/docs/reference/styles/style-selector-syntax

**Reply by maxkatz6** [reply]
> but I don't understand how this applies to a control and/or a specific state like normal, hovered, focused, pressed, disabled.

ControlThemes documentation also has some info on how to manage different states. It's still the same ControlTheme, but you add an additional nested styles per each state:
https://docs.avaloniaui.net/docs/basics/user-interface/styling/control-themes#interaction-in-control-themes

```xaml
  <ControlTheme TargetType="{x:Type Button}"
                x:Key="CustomTheme">
    <Setter Property="Template">
      <ControlTemplate>
           <!-- same as before ... -->
      </ControlTemplate>
    </Setter>

    <Style Selector="^:pointerover /template/ Border">
      <Setter Property="Opacity" Value="0.9" />
    </Style>
    <Style Selector="^:pressed /template/ Border">
      <Setter Property="Opacity" Value="0.8" />
    </Style>
  </ControlTheme>
```

**Reply by maxkatz6** [reply]
I used Opacity to visualize different states, but if you add x:Name on your Border nodes, you can access them and change colors.

**Reply by Shadowblitz16** [reply]
Thankyou so much for explaining!

So what is the `PART_ContentPresenter` string mean?
And is the "^:" part necessary on state selection?

**Reply by maxkatz6** [reply]
> So what is the PART_ContentPresenter string mean?

That's a name of the template part. ContentControls (including Button) look for this part of the template by name, so they can setup components tree properly. 

> And is the "^:" part necessary on state selection?

"^" part is necessary, because this selector is relative to the control theme and not global. It's essentially required on any nested styles https://docs.avaloniaui.net/docs/reference/styles/style-selector-syntax#nesting

**Reply by Shadowblitz16** [reply]
@maxkatz6 how do I access my named Borders and set it's BorderBrush?
```xml
<ResourceDictionary xmlns="https://github.com/avaloniaui"
                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">
    <!-- Add Resources Here -->
    <Design.PreviewWith>
        <Button Content="Hello World!"
                Theme="{StaticResource OldSchool}"
                VerticalContentAlignment="Center"/>
    </Design.PreviewWith>
    
    <FontFamily x:Key="Sans">/Themes/OldSchool/Fonts/SansSerif.ttf</FontFamily>
    <ControlTheme TargetType="{x:Type Button}" x:Key="OldSchool">
        <Setter Property="Template">
            <ControlTemplate>
                <Border x:Name="OuterBorderMin" BorderThickness="1,1,0,0" BorderBrush="#d79773">
                    <Border x:Name="OuterBorderMax" BorderThickness="0,0,1,1" BorderBrush="#4f2b13">
                        <Border x:Name="InnerBorderMin" BorderThickness="1,1,0,0" BorderBrush="#ffe3c3">
                            <Border x:Name="InnerBorderMax" BorderThickness="0,0,1,1" BorderBrush="#8b573b">
                                <Border BorderThickness="3,3,3,3" BorderBrush="#d79773" Background="#d79773">
                                    <ContentPresenter x:Name="PART_ContentPresenter"
                                          Content="{TemplateBinding Content}"
                                          Margin="{TemplateBinding Padding}"
                                          Foreground="Black"
                                          FontFamily="Sans"
                                          FontSize="9"
                                          HorizontalContentAlignment="{TemplateBinding HorizontalContentAlignment}"
                                          VerticalContentAlignment="{TemplateBinding VerticalContentAlignment}"  />
                                </Border>
                            </Border>
                        </Border> 
                    </Border>
                </Border>
            </ControlTemplate>
        </Setter>
        <Style Selector="^:pressed>
            <Setter Property="Opacity" Value="0.8" />
        </Style>
    </ControlTheme>
</ResourceDictionary>
```

Also it would be nice if I could export the colors to the root of the theme.

**Reply by maxkatz6** [reply]
@Shadowblitz16 you need template and name selectors

```xml
    <Style Selector="^:pointerover /template/ Border#InnerBorderMin">
      <Setter Property="Opacity" Value="0.9" />
    </Style>
```

**Reply by maxkatz6** [reply]
> Also it would be nice if I could export the colors to the root of the theme.

Yeah, you can do that. Define colors in the root resource dictionary. And use DynamicResource to reference them by key where its used.
https://docs.avaloniaui.net/docs/guides/styles-and-resources/resources

**Reply by Shadowblitz16** [reply]
> > Also it would be nice if I could export the colors to the root of the theme.
> 
> Yeah, you can do that. Define colors in the root resource dictionary. And use DynamicResource to reference them by key where its used. https://docs.avaloniaui.net/docs/guides/styles-and-resources/resources

How do I have a theme selector select multipule objects?
so instead of this...
```xml
        <Style Selector="^:pressed/template/Border#InnerBorderMin">
            <Setter Property="BorderBrush" Value="#ffe3c3"/>
        </Style>
        <Style Selector="^:pressed/template/Border#InnerBorderMax">
            <Setter Property="BorderBrush" Value="#ffe3c3"/>
        </Style>
```

is something like this possible?
```xml
        <Style Selector="^:is:pressed/template/>
            <Setter Property="Border#InnerBorderMin/BorderBrush" Value="{DynamicResource Shade3}"/>
            <Setter Property="Border#InnerBorderMax/BorderBrush" Value="{DynamicResource Shade3}"/>
            <Setter Property="Border#OuterBorderMin/BorderBrush" Value="{DynamicResource Shade3}"/>
            <Setter Property="Border#OuterBorderMax/BorderBrush" Value="{DynamicResource Shade3}"/>
        </Style>
```

**Reply by maxkatz6** [reply]
No, each selector needs to target a single control.
There is a feature proposal to support what you want though https://github.com/AvaloniaUI/Avalonia/issues/11537

**Reply by Shadowblitz16** [reply]
Wait themes only apply to one control at a time?
I tried using my whole app as a preview but nothing is changing

```xml
<ResourceDictionary xmlns="https://github.com/avaloniaui"
                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
                    xmlns:v="clr-namespace:MarioKit">
    <!-- Add Resources Here -->
    
    <Design.PreviewWith>
        <v:MainView Width="640" Height="280" Theme="{StaticResource Classic}"/>
    </Design.PreviewWith>
    
    <ControlTheme TargetType="{x:Type v:MainView}" x:Key="Classic">
        <Setter Property="Background" Value="#8f6327"/>
        <Setter Property="MinHeight" Value="24"/>
        <Setter Property="Template">
        <ControlTemplate>
            <Border BorderThickness="1,1,0,0"  BorderBrush="#d79773">
                <Border BorderThickness="0,0,1,1" BorderBrush="#4f2b13">
                    <Border BorderThickness="1,1,0,0" BorderBrush="#ffe3c3">
                        <Border BorderThickness="0,0,1,1" BorderBrush="#8b573b">
                            <ContentPresenter x:Name="PART_ContentPresenter"
                                              Background="{TemplateBinding Background}"
                                              Content="{TemplateBinding Content}"
                                              Margin="{TemplateBinding Padding}"
                                              HorizontalAlignment="{TemplateBinding HorizontalContentAlignment}"
                                              VerticalAlignment="{TemplateBinding VerticalContentAlignment}"/>
                        </Border>
                    </Border>
                </Border>
            </Border>
        </ControlTemplate>
        </Setter>
    </ControlTheme>
</ResourceDictionary>

```

If this is the case then that's alot of boilerplate.

**Reply by maxkatz6** [reply]
> Wait themes only apply to one control at a time?

Not sure if I understood this question. Single ControlTheme can be applied to many controls.

> I tried using my whole app as a preview but nothing is changing

`PreviewWith` is a **preview** only thing, just to visualize your file while you are editing, and used with Avalonia Extension previewer.

Once this ResourceDictionary is included in the app, you only need to write `<Button Theme="{StaticResource Classic}" />`.
Or, if you want to make it global (i.e., for all buttons), then ControlTheme x:Key should be `x:Key="{x:Type Button}"`. 
Please, double check control themes documentation https://docs.avaloniaui.net/docs/basics/user-interface/styling/control-themes.

**Reply by maxkatz6** [reply]
> <ControlTheme TargetType="{x:Type v:MainView}" x:Key="Classic">

I don't think you really want to create control theme for the main view.

**Reply by Shadowblitz16** [reply]
> > 
> 
> I don't think you really want to create control theme for the main view.

I want to create a global theme in which all controls are styled

**Reply by Shadowblitz16** [reply]
@maxkatz6 I still doesn't understand Themes but Styles seem to be much easier.
Is there a way to create a custom BorderBrush for control that draws a 3d border like that?

**Comment by Shadowblitz16** [comment]
@maxkatz6 how do I theme multiple controls using my mainview as a preview?
```axaml
<ResourceDictionary xmlns="https://github.com/avaloniaui"
                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
                    xmlns:v="using:SMBX.Views">
    <Design.PreviewWith>
         <v:MainView Width="800" Height="450" Theme="{StaticResource Slick}"/>
    </Design.PreviewWith>
    
    <ControlTheme TargetType="{x:Type v:MainView}" x:Key="Slick">
        <Styles>
            <Style Selector="Button">
                <Setter Property="Padding" Value="3"/>
                <Setter Property="CornerRadius" Value="3"/>
            </Style>
        </Styles>
    </ControlTheme>
</ResourceDictionary>

```

**Reply by maxkatz6** [reply]
Single ControlTheme is applied on the single control. It cannot leak to nested elements.
Existing built-in and third-party theme packs define multiple ControlThemes per each control type (typically), and combined them with ResourceDictionary.

**Reply by maxkatz6** [reply]
You can use MainView as a preview. But it would be more like:
```
<ResourceDictionary xmlns="https://github.com/avaloniaui"
                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
                    xmlns:v="using:SMBX.Views">
    <Design.PreviewWith>
         <v:MainView Width="800" Height="450" />
    </Design.PreviewWith>
    
    <ControlTheme TargetType="{x:Type Button}" x:Key="{x:Type Button}">
          <Setter Property="Padding" Value="3"/>
          <Setter Property="CornerRadius" Value="3"/>
    </ControlTheme>
</ResourceDictionary>
```

**Reply by Shadowblitz16** [reply]
> You can use MainView as a preview. But it would be more like:
> 
> ```
> <ResourceDictionary xmlns="https://github.com/avaloniaui"
>                     xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
>                     xmlns:v="using:SMBX.Views">
>     <Design.PreviewWith>
>          <v:MainView Width="800" Height="450" />
>     </Design.PreviewWith>
>     
>     <ControlTheme TargetType="{x:Type Button}" x:Key="{x:Type Button}">
>           <Setter Property="Padding" Value="3"/>
>           <Setter Property="CornerRadius" Value="3"/>
>     </ControlTheme>
> </ResourceDictionary>
> ```

ty so much! I will look into this more.
A guy on discord explained the avalonia theming system.

The reason it confused me is because coming from the Godot Engine you might have a collection of styles for a control type like this...
- normal,
- pressed,
- hovered,
- focused,
- disabled,

And then a theme was a collection of control types.
- Panel styles
- Button styles,
- CheckBox styles

He explained to me a control theme in avalonia is more like a collection of styles and a resource dictionary is more like a collection of control types.

**Reply by maxkatz6** [reply]
Yeah. `ControlTheme` specifically targets a specific control in Avalonia. While something like `FluentTheme` is a collection of control themes with fluent design, it will use resource dictionary internally to actually store these control themes.

**Reply by Shadowblitz16** [reply]
> <Design.PreviewWith>
>          <v:MainView Width="800" Height="450" />
>     </Design.PreviewWith>

It's not updating in the designer.
```axaml
<ResourceDictionary xmlns="https://github.com/avaloniaui"
                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
                    xmlns:v="using:SMBX.Views">
    <Design.PreviewWith>
        <v:EditorView Width="800" Height="450" />
    </Design.PreviewWith>
    
    <ControlTheme TargetType="{x:Type Button}" x:Key="{x:Type Button}">
        
        <Style Selector="^">
            <Setter Property="Padding" Value="3"/>
            <Setter Property="CornerRadius" Value="3"/> 
            <Setter Property="Background" Value="#505050"/>
        </Style>
        <Style Selector="^:pointerover">
            <Setter Property="Padding" Value="3"/>
            <Setter Property="CornerRadius" Value="3"/> 
            <Setter Property="Background" Value="#606060"/>
        </Style>
        <Style Selector="^:pressed">
            <Setter Property="Padding" Value="3"/>
            <Setter Property="CornerRadius" Value="3"/> 
            <Setter Property="Background" Value="#404040"/>
        </Style>
    </ControlTheme>
    
</ResourceDictionary>

```

Any reason why?

Edit: Also idk how to include my Slick.axaml in my App.axaml(Program.axaml)

This says it doesn't exist....
```axaml
<Application xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             x:Class="SMBX.Program"
             
             RequestedThemeVariant="Dark">
             <!-- "Default" ThemeVariant follows system theme variant. "Dark" or "Light" are other available options. -->
  
    <Application.Styles>
        <Slick/>
    </Application.Styles>
</Application>
```

![image](https://github.com/user-attachments/assets/ab5b45ab-9a3c-48ad-85ff-f5d3b45e77a7)

**Reply by stevemonaco** [reply]
That `ControlTheme` has no `ControlTemplate`. Check out the Fluent control's XAML for `Button`: https://github.com/AvaloniaUI/Avalonia/blob/411b82f221f2863cf4de6f7af371cf8b9be3fd9b/src/Avalonia.Themes.Fluent/Controls/Button.xaml

While you don't need all of this starting out, you do need the `ControlTemplate` since you aren't using `BasedOn`. A `TemplatedControl` doesn't do any drawing itself -- instead it's composed of child controls. No `ControlTemplate` in this case, no display. You can test it out with something simpler to begin with, like a `<TextBlock Text="12345" />`, before moving onto the `ContentPresenter`.

To include like `<FluentTheme />`, you need to create a `Styles` like I presume you did. You then need to assign it an `x:Class` and give it an empty code-behind class which you'll need to do manually. Then you just need the `xmlns` in the XAML file and specify the prefix. You can follow the Fluent theme if you get lost. I didn't bother until our project was about 20 definitions deep.

**Reply by Shadowblitz16** [reply]
> That `ControlTheme` has no `ControlTemplate`. Check out the Fluent control's XAML for `Button`: https://github.com/AvaloniaUI/Avalonia/blob/411b82f221f2863cf4de6f7af371cf8b9be3fd9b/src/Avalonia.Themes.Fluent/Controls/Button.xaml
> 
> While you don't need all of this starting out, you do need the `ControlTemplate` since you aren't using `BasedOn`. A `TemplatedControl` doesn't do any drawing itself -- instead it's composed of child controls. No `ControlTemplate` in this case, no display. You can test it out with something simpler to begin with, like a `<TextBlock Text="12345" />`, before moving onto the `ContentPresenter`.
> 
> To include like `<FluentTheme />`, you need to create a `Styles` like I presume you did. You then need to assign it an `x:Class` and give it an empty code-behind class which you'll need to do manually. Then you just need the `xmlns` in the XAML file and specify the prefix. You can follow the Fluent theme if you get lost. I didn't bother until our project was about 20 definitions deep.

Right now I don't want to define my own control theme only override how simple theme looks.
Is this not possible?

Also it still doesn't work...
```axaml
<ResourceDictionary xmlns="https://github.com/avaloniaui"
                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
                    xmlns:v="using:SMBX.Views">
    <Design.PreviewWith>
        <v:EditorView Width="800" Height="450" />
    </Design.PreviewWith>
    
    <ControlTheme TargetType="{x:Type ToggleButton}" x:Key="ToggleButton">
        <Setter Property="Background" Value="{DynamicResource ThemeControlMidBrush}" />
        <Setter Property="Foreground" Value="{DynamicResource ThemeForegroundBrush}" />
        <Setter Property="BorderThickness" Value="{DynamicResource ThemeBorderThickness}" />
        <Setter Property="BorderBrush" Value="{DynamicResource ThemeBorderLowBrush}" />
        <Setter Property="HorizontalAlignment" Value="Center" />
        <Setter Property="VerticalAlignment" Value="Center" />
        <Setter Property="Padding" Value="3" />
        <Setter Property="Template">
            <ControlTemplate>
                <ContentPresenter Name ="PART_ContentPresenter"
                                  HorizontalAlignment="{TemplateBinding HorizontalAlignment}"
                                  VerticalAlignment="{TemplateBinding VerticalAlignment}"
                                  Margin="{TemplateBinding Margin}"
                                  Padding="{TemplateBinding Padding}"
                                  Background="{TemplateBinding Background}"
                                  Foreground="{TemplateBinding Foreground}"
                                  BorderBrush="{TemplateBinding BorderBrush}"
                                  BorderThickness="{TemplateBinding BorderThickness}"
                                  CornerRadius="{TemplateBinding CornerRadius}"
                                  Content="{TemplateBinding Content}"
                                  ContentTemplate="{TemplateBinding ContentTemplate}"
                                  TextElement.Foreground="{TemplateBinding Foreground}">
                </ContentPresenter>
            </ControlTemplate>
        </Setter>
        

        <Style Selector="^:pointerover/template/ContentPresenter#PART_ContentPresenter">
            <Setter Property="Padding" Value="3"/>
            <Setter Property="CornerRadius" Value="3"/> 
            <Setter Property="Background" Value="#606060"/>
        </Style>
        <Style Selector="^:pressed/template/ContentPresenter#PART_ContentPresenter">
            <Setter Property="Padding" Value="3"/>
            <Setter Property="CornerRadius" Value="3"/> 
            <Setter Property="Background" Value="#404040"/>
        </Style>
    </ControlTheme>
    
</ResourceDictionary>

```

**Reply by stevemonaco** [reply]
Keep in mind you're using a `ToggleButton` now. Please use proper spacing in areas like `^:pointerover/template/ContentPresenter#PART_ContentPresenter` as it matters, IIRC.  I would strip away the `<v:EditorView ... />` until you get it working. Make like a 400x400 `Panel` with the `Button` or `ToggleButton` inside. Get that working, then revert. Don't start off with many pieces that can potentially break if you can avoid it.

If you're adjusting the Simple theme, then you still need to include the simple theme in App.axaml (before your `<SlickTheme />` inclusion).

Whether, you need to re-template depends on how deeply you're trying to customize. If you're only changing a few preexisting brushes (that aren't shared), then you might not even need a `ControlTheme`. Once you start trying to change hardcoded values (not from resources) or more importantly, changing the control types or ordering in the `ControlTemplate`, then you're in the territory of needing to create a new/derived `ControlTheme`.

**Reply by timunie** [reply]
> Right now I don't want to define my own control theme only override how simple theme looks.
> Is this not possible?

It is possible if you include SimpleTheme and use `BasedOn="{StaticResource {x:Type YourControlToOverride}}"`.

**Reply by Shadowblitz16** [reply]
> Keep in mind you're using a `ToggleButton` now. Please use proper spacing in areas like `^:pointerover/template/ContentPresenter#PART_ContentPresenter` as it matters, IIRC. I would strip away the `<v:EditorView ... />` until you get it working. Make like a 400x400 `Panel` with the `Button` or `ToggleButton` inside. Get that working, then revert. Don't start off with many pieces that can potentially break if you can avoid it.
> 
> If you're adjusting the Simple theme, then you still need to include the simple theme in App.axaml (before your `<SlickTheme />` inclusion).
> 
> Whether, you need to re-template depends on how deeply you're trying to customize. If you're only changing a few preexisting brushes (that aren't shared), then you might not even need a `ControlTheme`. Once you start trying to change hardcoded values (not from resources) or more importantly, changing the control types or ordering in the `ControlTemplate`, then you're in the territory of needing to create a new/derived `ControlTheme`.

That's the issue though, I can get it to display with one button but I can't get it to display inside another control.

Edit: OH, I think it's because I have not using the Theme property here...
```axaml
    <Design.PreviewWith>
        <v:EditorView Width="800" Height="450" Theme="" />
    </Design.PreviewWith>
```

But how do I actually use the resource dictionary for the design?
A control theme is only for a single control.
I want to view the whole resource dictionary when creating it.

> BasedOn="{StaticResource {x:Type YourControlToOverride}}"

Doesn't seem to work.
`{x:Type SimpleTheme}` is yellow highlighted with the warning that the resource is not found.
You sure there is no typo?

```axaml
<ResourceDictionary xmlns="https://github.com/avaloniaui"
                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
                    xmlns:v="using:SMBX.Views">
    <Design.PreviewWith>
        <v:EditorView Width="800" Height="450" />
    </Design.PreviewWith>
    
    <ControlTheme TargetType="{x:Type ToggleButton}" x:Key="ToggleButton" BasedOn="{StaticResource {x:Type SimpleTheme}}">
        <Setter Property="Background" Value="{DynamicResource ThemeControlMidBrush}" />
        <Setter Property="Foreground" Value="{DynamicResource ThemeForegroundBrush}" />
...
```

**Reply by timunie** [reply]
```xml
    <ControlTheme TargetType="{x:Type ToggleButton}" x:Key="ToggleButton" BasedOn="{StaticResource {x:Type ToggleButton}}">
        <Setter Property="Background" Value="{DynamicResource ThemeControlMidBrush}" />
        <Setter Property="Foreground" Value="{DynamicResource ThemeForegroundBrush}" />

**Reply by Shadowblitz16** [reply]
> ```
>     <ControlTheme TargetType="{x:Type ToggleButton}" x:Key="ToggleButton" BasedOn="{StaticResource {x:Type ToggleButton}}">
>         <Setter Property="Background" Value="{DynamicResource ThemeControlMidBrush}" />
>         <Setter Property="Foreground" Value="{DynamicResource ThemeForegroundBrush}" />
> ```

1. How do I name a resource dictionary so I can include it in `<Application.Styles>`?
2. How do I include other control themes in my styles resource dictionary?
